### ISSUE TITLE
Support for sphinx style docstrings

### ISSUE THREAD

**Issue by machow** [mkdocstrings/griffe#193]
**Is your feature request related to a problem? Please describe.**

In sphinx, a special [comment-based format](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoproperty) is available for creating docstrings for attributes:

```python
#: a docstring for a
a = 1

#: a docstring for b
#: that spans multiple lines
b

c    #: a docstring on the same line
```

**Describe the solution you'd like**

What do you think of attempting to parse these docstrings? It seems like currently griffe looks at the next AST node, to see if it can add a docstring for attributes ([code here](https://github.com/mkdocstrings/griffe/blob/main/src/griffe/agents/visitor.py#L608)).

I wonder if we could extend it to do the following (edited for clarity):

* **prefer AST docstrings**: if next node is a string, use that (current approach). otherwise,
* **upward search**: 
  - if the previous node is not on the same line, or the one above,
  - if de-dented line above starts with `#:`, parse docstring from that line upwards (until no more `#:`).
  - otherwise,
* **rightward search**:
  - if next node is not on the same line
  - look on the same line as node for `#:`

I'm happy to take a crack at it / incorporate any feedback!

**Describe alternatives you've considered**

Changing docstrings to use the `"""` format:

```python
a
"""a docstring for a"""

b
"""
a docstring for b
that spans multiple lines
"""
```

This is nice, in that it keeps things in the python AST. The downside is it requires older libraries to update their approach, and can read a bit funky.

**Comment by pawamoy** [comment]
Unfortunately, this will be hard to implement, and IMO is not worth it, for the following reasons:

1. we indeed use `ast` to parse source code
2. `ast` does not keep comments
3. so it means we'd have to also tokenize the code with `tokenize`, greatly reducing perfs (at least when enabled)
4. even with tokens, it will be hard to identify on which object to attach the docstrings found in comments
5. these comments are a Sphinx-thing, not a standard (see https://peps.python.org/pep-0257/#what-is-a-docstring, which points anyway to the (rejected) https://peps.python.org/pep-0258/#attribute-docstrings)

Even if someone was willing to work on this and send a PR, I myself am not willing to review such a big feature, even less maintain it :confused:

**Comment by machow** [comment]
Thanks for your quick response, and background on the issue. It's clear you've put a lot of thought into this scenario!

I had assumed `LinesCollection` holds the lines for the source code, which would allow fast checks / extraction (e.g. does this line start with `#:`?). Is it used in a different way?

I was hoping that if..

* the AST nodes tell us where they start and stop (e.g. line number)
* LinesCollection holds lines of source code

Simple extraction of these comments might be relatively simple. But I'm not too familiar with everything that's available at the time docstrings are extracted by the agent, so totally understand if it is a much bigger lift.

**Comment by pawamoy** [comment]
Oh. OH.

This is a moment where I realize again that external points of view are *so* valuable.

You're completely right, the lines collection could make this *way* easier. I didn't even think of that.

However getting inline comments would still be not trivial without a tokenizer:

```python
do(thing, "  #: this is not a comment!")
```

...but we can start with a simple approach that solves 99% of the cases.

**Comment by machow** [comment]
Ahhh, teamwork!

I think that--if `do(...)` is the node we're interested in, then if we used the three rules in my initial post (which I edited a bit for clarity; thoroughness..!), we wouldn't extract a comment from this code:

```python
do(thing, "  #: this is not a comment!")
```

Since the `do(...)` node will tell you it ends at `)` in the source.

The **rightward search rule** would look to the right of that node...

```python
do(thing, "  #: this is not a comment!") #: this is a comment
```

## More cases

### Upward search (early termination)

```python
a = 1  #: I am a comment
b = 2
```

When searching for docs for `b`, we wouldn't look upward because the previous node is on the line above.

Similarly, it wouldn't search upward for `b` here...

```python
#: some comment for a?
a = 1; b = 2
```

### Rightward search (multiline node)

```python
b = int(
    "1"
) #: some comment
```

### Rightward search (multiple nodes per line)

```python
a = 1; b = 2;     #: a comment for b
```

**Comment by pawamoy** [comment]
My example was not good indeed.

Maybe this one is better:

```python
a = "  #: not a comment"
```

And you're right again, we can probably use the ending column info from the node to know where to look at for a comment :+1:

**Comment by pushfoo** [comment]
TL;DR:
1. OP's line link seems outdated due to refactoring
2. Could this be implemented in phases to ship faster?
3. Is OP still interested in this?

### 1. OP's Link is Outdated

TL;DR: The relevant lines now seem to be [here](https://github.com/mkdocstrings/griffe/blob/58b17425bace2072664fb654ba1077fb404b92dd/src/griffe/agents/visitor.py#L543-L568)

There have been a lot of changes to `visitor.py` since this ticket was created. The link in OP currently points to a docstring in an augmented assignment (`+=`, etc) handler.

### 2. Could we implement this in phases to ship faster?

TL;DR: Could we start by supporting multi-line only?

* Partial Sphinx docstring support would be better than nothing
* Documenting limitations could compensate for imperfect support

#### Why Multi-Line?

* It'd make a single line above an attribute the simplest case (n=1)
* Above-the-attribute is the only style I've seen recently in public code bases

N=2, but there seem to be 200+ usages in each:

[pyglet]: https://github.com/pyglet/pyglet
[Arcade]: https://github.com/pythonarcade/arcade

| Library GitHub Page | TL;DR                                                              |
|-----------------------------|-------------------------------------------------------------|
| [pyglet][]                    | Pure Python ctypes multimedia & windowing | 
| [Arcade][]                  | Game framework built on pyglet                     |

The other use cases could be added afterward if there's demand / data to support them.

### 3. Is OP still interested?

TL;DR: @machow Are you still interested in this ticket?

If not, I want to do the following before promising anything:

1. Read the code more thoroughly
2. Wait for other Arcade contributors to respond to my ping about this ticket in Discord


### Bonus: Arcade's Use Case

TL;DR: Fix tech debt in our API doc generation

Arcade was once much smaller. At the time, using regex to parse code and generate Sphinx autodoc directives worked okay. As of our 3.0 dev previews, it's causing problems.

**Comment by pawamoy** [comment]
This is now supported thanks to the Griffe Sphinx extension, available to sponsors :slightly_smiling_face: See https://mkdocstrings.github.io/griffe-sphinx/!
